### ISSUE TITLE
getDiscriminatorModelForClass doesn't respect the model's existingConnection option

### ISSUE THREAD

**Issue by CasperLinden2** [typegoose/typegoose#789]
## Versions

- System: linux 
- NodeJS: v18.13.0
- Typescript: 4.9.4

- Compiler / Transpiler: tsc
- Typegoose(NPM): 10.1.0
- mongoose: 6.9.0
- mongodb: 4.13.0

## What is the Problem?

I would expect getDiscriminatorModelForClass to use the same connection that the base model uses. But that doesn't happen, even if you explicitly specify it.

## Code Example

```ts
const db = await mongoose.createConnection('mongodb://127.0.0.1:27017/database1').asPromise();

const model1 = getModelForClass(ModelOne, {
    existingConnection: db.useDB('database2')
});

// Model 1 works fine

const model2 = getDiscriminatorModelForClass(ModelOne, ModelTwo, {
    existingConnection: db.useDB('database2')
});

// model2 is using database1

```

## Do you know *why* it happens?

*no*

---

**Comment by hasezoey** [comment]
> I would expect getDiscriminatorModelForClass to use the same connection that the base model uses

it should, from what i know

> even if you explicitly specify it.

from what i know `existingMongoose` and `existingConnection` cannot be set differently for a discriminator

> `const model2 = getDiscriminatorModelForClass(ModelOne, ModelTwo, {`

are you sure you want a discriminator on `ModelOne` instead of `model1`?

**Comment by CasperLinden2** [comment]
@hasezoey Sorry, that was my lazy translation from the real source.

I've dug deeper and found that it was my bug, **however**, it would be super helpful if this spat out a warning ;)

```
import {getDiscriminatorModelForClass, getModelForClass, prop} from "@typegoose/typegoose";
import mongoose from "mongoose";

export class ModelOne
{
    @prop()
    public var2: string;
}

export class ModelTwo extends ModelOne
{
    @prop()
    public var3: string;
}


async function test()
{
    const db = await mongoose.createConnection('mongodb://127.0.0.1:27017/database1', {}).asPromise();

    // To reproduce the issue we have to first get the model from the first db
    getModelForClass(ModelTwo, {
        existingConnection: db
    });

    const newDB = db.useDb('database2');
    const model2 = getModelForClass(ModelOne, {
        existingConnection: newDB
    });
    const model3 = getDiscriminatorModelForClass(model2, ModelTwo, {
        existingConnection: newDB
    });

    const test = new model3();
    test.var3 = 'test';
    await test.save();

    // Model3 will be created on database1
}
test().catch((e) =>
{
    console.error(e);
})


```

**Comment by hasezoey** [comment]
with 0599ef299901e5eb4a47e64fea298a802f25d9c5 there is documentation in `getDiscriminatorModelForClass` noting that `existing*` options will not be used from the discriminator, it also adds a runtime warning as `W002`.
this change will be in version `10.1.1`

**Comment by CasperLinden2** [comment]
Actually, I don't think the document change covers this entirely.

I think we also need `Warn if Model was already declared on a different connection`

**Comment by hasezoey** [comment]
> I think the issue is that you can't define the same model on multiple connections/databases.

no that is possible, that is a different issue - which is already documented as [Models with same name](https://typegoose.github.io/typegoose/docs/guides/advanced/models-with-same-name)
TL;DR: it is because typegoose does its own model / class caching, which is based on model name, to be able to retrieve a class for a registered model

or if this is not what you mean, please clarify

**Comment by CasperLinden2** [comment]
Well, from my example script

I first get ModelTwo on **database1**

```
getModelForClass(ModelTwo, {
        existingConnection: db
});
```

Then a subsequent attempt to get ModelTwo on **database2** results in it ending up on **database1** despite it deriving from a model which is successfully established on database2

```
 const model2 = getDiscriminatorModelForClass(model1, ModelTwo, {
        existingConnection: newDB
    });
```

**Comment by CasperLinden2** [comment]
So actually because of this double-initialisation it results in the opposite behaviour than is documented by your changes

**Comment by hasezoey** [comment]
> opposite behaviour than is documented by your changes

the documented behavior is not for double defining the same model name

i will consider adding a way to disable caching so this can be allowed (while disabling the getting from cache)

**Comment by hasezoey** [comment]
released typegoose [`v10.2.0-beta.1`](https://github.com/typegoose/typegoose/releases/tag/v10.2.0-beta.1) which includes a way to disable caching, please test it and report if this solves your issue

for now the beta documentation can be found at [Github Models with same name: Disable Caching](https://github.com/typegoose/typegoose/blob/v10.2.0-beta.1/docs/guides/advanced/models-with-same-name.md#disable-caching)

**Comment by CasperLinden2** [comment]
Thanks! Could this be an option that could be passed to `getModelForClass` and `getDiscriminatorModelForClass` so we don't have to disable caching globally?

**Comment by hasezoey** [comment]
locally disabled cache is added with 78ac3bce1536cf667d454fd02810035c6758901a, which will be in [`10.2.0-beta.2`](https://github.com/typegoose/typegoose/releases/tag/v10.2.0-beta.2)

please test if this is properly working

for now the beta documentation can be found at [Github Models with same name: Disable Caching](https://github.com/typegoose/typegoose/blob/v10.2.0-beta.2/docs/guides/advanced/models-with-same-name.mdx#disable-caching) AND [Github modelOptions: disableCaching](https://github.com/typegoose/typegoose/blob/v10.2.0-beta.2/docs/api/decorators/modelOptions.md#disablecaching)

(also the global option has been renamed from `disableCaching` to `disableGlobalCaching`)
